File: db/resources.go
package db

import (
	"fmt"
	"sort"

	watcherlib "github.com/mturley/watcher"
	wdb "github.com/mturley/watcher/db"
)

// ResourceRelationship represents a hierarchical relationship between resources.
type ResourceRelationship struct {
	ID           string
	ChildType    string
	ChildID      string
	ChildURL     *string
	ParentType   string
	ParentID     string
	ParentURL    *string
	Relationship string
	Source       string
	CreatedAt    string
}

// LinkResources records a hierarchical relationship between two resources in
// the watcher library's watcher_resource_relationships table.
func (db *DB) LinkResources(r ResourceRelationship) error {
	child := watcherlib.Resource{Type: r.ChildType, ID: r.ChildID, URL: deref(r.ChildURL)}
	parent := watcherlib.Resource{Type: r.ParentType, ID: r.ParentID, URL: deref(r.ParentURL)}
	if err := wdb.LinkResources(db.conn, child, parent, r.Relationship, r.Source); err != nil {
		return fmt.Errorf("failed to link resources: %w", err)
	}
	return nil
}

// FindRelatedSessions finds sessions that share direct resource subscriptions OR subscribe to resources with the same parent.
// Excludes the given session and archived sessions. Returns sessions ordered by last_active DESC.
//
// Subscriptions now live in the watcher library's watcher_subscriptions
// table (see db/watcher_bridge.go), so this walks the library via
// SubscribersOf instead of joining handler's old subscriptions table.
// Resource relationships also live in the library's
// watcher_resource_relationships table; sibling lookups go through
// wdb.SiblingResources.
func (db *DB) FindRelatedSessions(sessionID string) ([]Session, error) {
	mine, err := wdb.ActiveSubscriptions(db.conn, handlerSubscriber(sessionID), false)
	if err != nil {
		return nil, fmt.Errorf("failed to list subscriptions for session %q: %w", sessionID, err)
	}
	if len(mine) == 0 {
		return nil, nil
	}

	related := map[string]bool{}
	addSubscribers := func(resourceType, resourceID string) error {
		subs, err := wdb.SubscribersOf(db.conn, resourceType, resourceID)
		if err != nil {
			return fmt.Errorf("failed to find subscribers of %s/%s: %w", resourceType, resourceID, err)
		}
		for _, s := range subs {
			if !isLiveSubscription(s) {
				continue
			}
			sid, ok := sessionIDFromSubscriber(s.Subscriber)
			if !ok || sid == sessionID {
				continue
			}
			related[sid] = true
		}
		return nil
	}

	for _, r := range mine {
		// Sessions subscribed to the same resource.
		if err := addSubscribers(r.Resource.Type, r.Resource.ID); err != nil {
			return nil, err
		}

		// Sessions subscribed to a sibling resource sharing a parent.
		siblings, err := wdb.SiblingResources(db.conn, watcherlib.Resource{Type: r.Resource.Type, ID: r.Resource.ID})
		if err != nil {
			return nil, fmt.Errorf("failed to find sibling resources of %s/%s: %w", r.Resource.Type, r.Resource.ID, err)
		}
		for _, sib := range siblings {
			if err := addSubscribers(sib.Type, sib.ID); err != nil {
				return nil, err
			}
		}
	}

	sessions := make([]Session, 0, len(related))
	for sid := range related {
		s, err := db.GetSession(sid)
		if err != nil {
			return nil, fmt.Errorf("failed to load related session %q: %w", sid, err)
		}
		if s == nil || s.Status == "archived" {
			continue
		}
		sessions = append(sessions, *s)
	}
	sort.Slice(sessions, func(i, j int) bool { return sessions[i].LastActive > sessions[j].LastActive })

	return sessions, nil
}

// ResourceHistory returns all events referencing a resource, ordered by ts
// DESC. ResourceHistory is resource-scoped (it joins on the resource's
// event-linkage table): only github/jira-sourced events ever carry a
// resource linkage, and those events live exclusively in the watcher
// library's watcher_events/watcher_event_resources tables post-migration —
// handler's own events table never has event_resources rows (agent/handler/
// web events aren't resource-scoped). So this reads watcher_events only.
func (db *DB) ResourceHistory(resourceType, resourceID string, limit int) ([]Event, error) {
	events, err := wdb.EventsForResource(db.conn, resourceType, resourceID)
	if err != nil {
		return nil, fmt.Errorf("failed to query resource history: %w", err)
	}

	// EventsForResource orders ascending; handler's ResourceHistory contract is ts DESC.
	out := make([]Event, 0, len(events))
	for i := len(events) - 1; i >= 0; i-- {
		out = append(out, eventFromWatcher(events[i]))
	}
	if limit > 0 && len(out) > limit {
		out = out[:limit]
	}
	return out, nil
}

// SessionsForResource returns all subscriptions (including deleted) for a resource.
func (db *DB) SessionsForResource(resourceType, resourceID string) ([]Subscription, error) {
	rows, err := wdb.SubscribersOf(db.conn, resourceType, resourceID)
	if err != nil {
		return nil, fmt.Errorf("failed to query sessions for resource: %w", err)
	}

	out := make([]Subscription, 0, len(rows))
	for _, r := range rows {
		sid, ok := sessionIDFromSubscriber(r.Subscriber)
		if !ok {
			continue // skip non-handler subscribers
		}
		out = append(out, subFromWatcher(sid, r))
	}
	return out, nil
}

File: src/Handler/Db/Resources.hs
-- | Port of db/resources.go: resource relationships and history.
module Handler.Db.Resources
  ( ResourceRelationship(..)
  , linkResources
  , findRelatedSessions
  , resourceHistory
  , sessionsForResource
  ) where

import Data.Text (Text)
import qualified Database.SQLite.Simple as SQL
import Database.SQLite.Simple.ToField (toField)

import Handler.Db (Db(..))
import Handler.Db.Events (Event(..))
import Handler.Db.Sessions (Session(..))
import Handler.Db.Subscriptions (Subscription(..))

-- | A hierarchical relationship between resources.
data ResourceRelationship = ResourceRelationship
  { relId        :: Text
  , childType    :: Text
  , childId      :: Text
  , childUrl     :: Maybe Text
  , parentType   :: Text
  , parentId     :: Text
  , parentUrl    :: Maybe Text
  , relationship :: Text
  , source       :: Text
  , createdAt    :: Text
  } deriving (Show, Eq)

linkResources :: Db -> ResourceRelationship -> IO ()
linkResources db r =
  SQL.execute db.conn
    "INSERT INTO resource_relationships (id, child_type, child_id, child_url, parent_type, parent_id, parent_url, relationship, source, created_at)\
    \ VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
    ( r.relId, r.childType, r.childId, r.childUrl, r.parentType
    , r.parentId, r.parentUrl, r.relationship, r.source, r.createdAt )

-- | Sessions sharing direct resource subscriptions OR subscribed to
-- resources with the same parent. Excludes the given session and archived
-- sessions; ordered by last_active DESC. (The Go version scans a reduced
-- column set; missing fields default to empty.)
-- | Wrapper for the reduced column set FindRelatedSessions selects
-- (sqlite-simple tuples stop at 10 fields).
newtype RelatedSessionRow = RelatedSessionRow Session

instance SQL.FromRow RelatedSessionRow where
  fromRow = do
    sessionId <- SQL.field
    harness <- SQL.field
    repo <- SQL.field
    branch <- SQL.field
    sessionName <- SQL.field
    pid <- SQL.field
    status <- SQL.field
    inboxMode <- SQL.field
    autoPollInterval <- SQL.field
    lastActive <- SQL.field
    registeredAt <- SQL.field
    jsonlPath <- SQL.field
    pure $ RelatedSessionRow Session
      { role = "", terminalType = "", terminalId = ""
      , cmuxWorkspaceId = "", cmuxWorkspaceName = "", cmuxWorkspaceColor = ""
      , lastPrompt = "", cwd = ""
      , ..
      }

findRelatedSessions :: Db -> Text -> IO [Session]
findRelatedSessions db sid = do
  rows <- SQL.query db.conn
    "SELECT DISTINCT\
    \  s.session_id, s.harness, s.repo, s.branch,\
    \  COALESCE(s.session_name, '') as session_name,\
    \  COALESCE(s.pid, 0) as pid,\
    \  s.status,\
    \  s.inbox_mode,\
    \  s.auto_poll_interval,\
    \  s.last_active, s.registered_at, s.jsonl_path\
    \ FROM sessions s\
    \ JOIN subscriptions sub ON s.session_id = sub.session_id AND sub.deleted_at IS NULL\
    \ WHERE s.session_id != ? AND s.status != 'archived'\
    \   AND (\
    \     (sub.resource_type, sub.resource_id) IN (\
    \       SELECT resource_type, resource_id\
    \       FROM subscriptions\
    \       WHERE session_id = ? AND deleted_at IS NULL\
    \     )\
    \     OR\
    \     EXISTS (\
    \       SELECT 1\
    \       FROM resource_relationships rr_other\
    \       JOIN resource_relationships rr_mine ON rr_mine.parent_type = rr_other.parent_type AND rr_mine.parent_id = rr_other.parent_id\
    \       JOIN subscriptions sub_mine ON sub_mine.resource_type = rr_mine.child_type AND sub_mine.resource_id = rr_mine.child_id\
    \       WHERE sub_mine.session_id = ? AND sub_mine.deleted_at IS NULL\
    \         AND rr_other.child_type = sub.resource_type AND rr_other.child_id = sub.resource_id\
    \         AND (rr_other.child_type != rr_mine.child_type OR rr_other.child_id != rr_mine.child_id)\
    \     )\
    \   )\
    \ ORDER BY s.last_active DESC"
    (sid, sid, sid)
  pure [ s | RelatedSessionRow s <- rows ]

-- | All events referencing a resource, ordered by ts DESC.
resourceHistory :: Db -> Text -> Text -> Int -> IO [Event]
resourceHistory db rType rId limit =
  SQL.query db.conn
    (SQL.Query ("SELECT DISTINCT e.id, e.ts, e.external_ts, e.source, e.session_id, e.type, e.title, e.body, e.author, e.author_type, e.broadcast, e.tags\
                \ FROM events e\
                \ JOIN event_resources er ON e.id = er.event_id\
                \ WHERE er.resource_type = ? AND er.resource_id = ?\
                \ ORDER BY e.ts DESC"
                <> (if limit > 0 then " LIMIT ?" else "")))
    ([toField rType, toField rId] ++ [toField limit | limit > 0])

-- | All subscriptions (including deleted) for a resource.
sessionsForResource :: Db -> Text -> Text -> IO [Subscription]
sessionsForResource db rType rId =
  SQL.query db.conn
    "SELECT id, session_id, resource_type, resource_id, resource_url, created_at, deleted_at\
    \ FROM subscriptions WHERE resource_type = ? AND resource_id = ?\
    \ ORDER BY created_at DESC"
    (rType, rId)
